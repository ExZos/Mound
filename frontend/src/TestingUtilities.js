import GeneralComponent from './components/GeneralComponent';

class TestingUtilities extends GeneralComponent {
  addUserSessionItem(userID, spaceID, pollID=undefined) {
    const user = {
      id: userID,
      name: 'name' + userID,
      space: spaceID,
      space_name: 'space' + spaceID,
      space_status: true,
      poll: pollID
    };

    this.addToSessionArrayItem('users', user);
  }
}

export default TestingUtilities;
