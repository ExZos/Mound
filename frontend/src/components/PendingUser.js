import React from 'react';

import GeneralComponent from './GeneralComponent';
import { server, api } from '../server';
import '../styles/pendingUser.css';

class PendingUser extends GeneralComponent {
  // TODO: create user from join poll either during vote creation or here
  constructor(props) {
    super(props);

    this.user = this.getSessionItem('users')[this.props.spaceID];

    this.state = {
      userCount: '',
      requiredVoteCount: '',
      positiveVoteCount: '',
      negativeVoteCount: '',
      remainingVoteCount: ''
    };
  }

  componentDidMount() {
    this.getPollResults();
    this.interval = setInterval(this.getPollResults, 1000);
  }

  // TODO: find better solution to fix default web navs
  //       should avoid full page reloads
  componentDidUpdate() {
    window.onpopstate = (e) => {
       window.location.reload(false);
    }
  }

  componentWillUnmount() {
    clearInterval(this.interval);
  }

  getPollResults = () => {
    // Get user count of space
    server.get(api.getJoinPollResults + this.user.poll + '/' + this.user.name)
      .then((res) => {
        const userCount = res.data['userCount'];
        const positiveVoteCount = res.data['positiveVoteCount'];
        const negativeVoteCount = res.data['negativeVoteCount'];

        this.setState({
          userCount: userCount,
          requiredVoteCount: userCount,
          positiveVoteCount: positiveVoteCount,
          negativeVoteCount: negativeVoteCount,
          remainingVoteCount: userCount - (positiveVoteCount + negativeVoteCount)
        });

        // Update user session item
        if(res.data['positiveVoteCount'] >= this.state.requiredVoteCount) {
          server.get(api.getUserInSpaceByName + this.user.space + '/' + this.user.name)
            .then((res) => {
              this.addToSessionArrayItem('users', res.data);

              this.props.updateState();
            });
        }
      });
  }

  displayStatusStatement = () => {
    if(this.state.positiveVoteCount < this.state.requiredVoteCount && this.state.remainingVoteCount === 0) {
      return 'rejected';
    }

    return 'pending';
  }

  displayResultingStatement = () => {
    if(this.state.positiveVoteCount < this.state.requiredVoteCount) {
      // Quota not met but still more users left to vote
      if(this.state.remainingVoteCount > 0) {
        return (
          <div>
            Some users have not yet voted.
          </div>
        );
      }

      // Join request rejected
      return (
        <div>
          All users have voted.<br />
          The required votes has not been met.<br />
          Your request has been rejected.
        </div>
      );
    }

    // Join request to be approved
    return (
      <div>
        The required votes has been met.<br />
        Your request will be approved shortly.
      </div>
    );
  }

  // TODO: display status rejected when applicable
  render() {
    return (
      <div id="pendingUser">
        <div className="statusStatement">
          Status: {this.displayStatusStatement()}
        </div>

        <div className="progressStatement">
          Progress: {this.state.positiveVoteCount}/{this.state.requiredVoteCount}
        </div>

        <br />

        <div className="resultingStatement">
          {this.displayResultingStatement()}
        </div>
      </div>
    );
  }
}

export default PendingUser;
