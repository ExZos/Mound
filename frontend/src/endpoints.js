export const endpoints = {
  messages: '/api/messages/',
  spaces: '/api/spaces/',
  users: '/api/users/',
  poll: '/api/polls/',
  votes: '/api/votes/',
  getSpaceByName: '/space/getByName/',
  getUserCountInSpaceForUser: '/space/getUserCountForUser/',
  getUserByName: '/user/getByName/',
  getUsersInSpace: '/user/getInSpace/',
  getUserInSpaceByName: '/user/getInSpaceByName/',
  getUsersInSpaceExceptName: '/user/getInSpaceExceptName/',
  createUserNApproveSpace: 'user/createNApproveSpace/',
  getMessagesInSpace: '/message/getInSpace/',
  getPendingPollsInSpace: '/poll/getPendingInSpace/',
  getPendingUnvotedPollsForUser: '/poll/getPendingUnvotedForUser/',
  getPendingJoinPollInSpaceByName: 'poll/getPendingJoinInSpaceByName/',
  getJoinPollResults: '/poll/getJoinResults/',
  createNameRelatedPoll: '/poll/createNameRelated/',
  getVotesForPoll: 'vote/getForPoll/',
  getPositiveVotesForPoll: 'vote/getPositivesForPoll/',
  createVoteNUpdatePoll: 'vote/createNUpdatePoll/',
}
