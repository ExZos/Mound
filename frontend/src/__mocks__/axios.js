import { endpoints as api } from '../endpoints';

const axios = {
  create: () => axios,
  get: (url) => {
    switch(url) {
      case api.getSpaceByName + "Headon's Floor/":
        return Promise.resolve({
          data: {
            id: 1,
            name: "Headon's Floor",
            status: true
          }
        });

      case api.getJoinPollResults + "1/" + "name1":
        return Promise.resolve({
          data: {
            'userCount': 10,
      			'positiveVoteCount': 5,
      			'negativeVoteCount': 3
          }
        });

      case api.getJoinPollResults + "2/" + "name2":
        return Promise.resolve({
          data: {
            'userCount': 3,
      			'positiveVoteCount': 3,
      			'negativeVoteCount': 0
          }
        });

      case api.getJoinPollResults + "3/" + "name3":
        return Promise.resolve({
          data: {
            'userCount': 7,
      			'positiveVoteCount': 2,
      			'negativeVoteCount': 5
          }
        });

      default:
        return Promise.reject();
    }
  }
};

export default axios;
