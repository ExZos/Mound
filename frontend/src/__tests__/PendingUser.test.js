import React from 'react';
import { shallow, mount, render } from 'enzyme';
import axios from 'axios';
import { createBrowserHistory } from 'history';

import PendingUser from '../components/PendingUser';
import TestingUtilities from '../TestingUtilities';
import { endpoints as api } from '../endpoints';

const history = createBrowserHistory();
const tutils = new TestingUtilities();

describe('PendingUser', () => {
  beforeEach(() => {
    tutils.addUserSessionItem(1, 1, 1);
    tutils.addUserSessionItem(2, 2, 2);
    tutils.addUserSessionItem(3, 3, 3);
  });

  it('should render pending correctly with props and session', async () => {
    const mockUpdateState = jest.fn();

    const spy = jest.spyOn(axios, 'get');
    const component = await shallow(<PendingUser spaceID={1} updateState={mockUpdateState} />)
    const statusStatement = component.find('div.statusStatement');
    const progressStatement = component.find('div.progressStatement');
    const resultingStatement = component.find('div.resultingStatement');

    expect(component).toMatchSnapshot();
    expect(spy).toHaveBeenCalled();
    expect(spy).toHaveBeenLastCalledWith(api.getJoinPollResults + '1/user1');
    expect(statusStatement.childAt(1).text()).toContain('pending');
    expect(progressStatement.childAt(1).text()).toEqual('5');
    expect(progressStatement.childAt(3).text()).toEqual('10');
    expect(resultingStatement.text()).toContain('not yet voted');
    expect(mockUpdateState).not.toHaveBeenCalled();
  });

  it('should render approved correctly with props and session', async () => {
    const mockUpdateState = jest.fn();

    const spy = jest.spyOn(axios, 'get');
    const component = await shallow(<PendingUser spaceID={2} updateState={mockUpdateState} />);
    const statusStatement = component.find('div.statusStatement');
    const progressStatement = component.find('div.progressStatement');
    const resultingStatement = component.find('div.resultingStatement');

    expect(component).toMatchSnapshot();
    expect(spy).toHaveBeenCalled();
    expect(spy).toHaveBeenLastCalledWith(api.getJoinPollResults + '2/user2');
    expect(statusStatement.childAt(1).text()).toContain('pending');
    expect(progressStatement.childAt(1).text()).toEqual('3');
    expect(progressStatement.childAt(3).text()).toEqual('3');
    expect(resultingStatement.text()).toContain('approved');
    expect(mockUpdateState).toHaveBeenCalled();
  });

  it('should render rejected correctly with props and session', async () => {
    const mockUpdateState = jest.fn();

    const spy = jest.spyOn(axios, 'get');
    const component = await shallow(<PendingUser spaceID={3} updateState={mockUpdateState} />)
    const statusStatement = component.find('div.statusStatement');
    const progressStatement = component.find('div.progressStatement');
    const resultingStatement = component.find('div.resultingStatement');

    expect(component).toMatchSnapshot();
    expect(spy).toHaveBeenCalled();
    expect(spy).toHaveBeenLastCalledWith(api.getJoinPollResults + '3/user3');
    expect(statusStatement.childAt(1).text()).toContain('rejected');
    expect(progressStatement.childAt(1).text()).toEqual('2');
    expect(progressStatement.childAt(3).text()).toEqual('7');
    expect(resultingStatement.text()).toContain('rejected');
    expect(mockUpdateState).not.toHaveBeenCalled();
  });
});
