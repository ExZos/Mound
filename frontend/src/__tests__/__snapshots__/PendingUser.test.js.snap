// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PendingUser should render approved correctly with props and session 1`] = `
<div
  id="pendingUser"
>
  <div
    className="statusStatement"
  >
    Status: 
    pending
  </div>
  <div
    className="progressStatement"
  >
    Progress: 
    3
    /
    3
  </div>
  <br />
  <div
    className="resultingStatement"
  >
    <div>
      The required votes has been met.
      <br />
      Your request will be approved shortly.
    </div>
  </div>
</div>
`;

exports[`PendingUser should render pending correctly with props and session 1`] = `
<div
  id="pendingUser"
>
  <div
    className="statusStatement"
  >
    Status: 
    pending
  </div>
  <div
    className="progressStatement"
  >
    Progress: 
    5
    /
    10
  </div>
  <br />
  <div
    className="resultingStatement"
  >
    <div>
      Some users have not yet voted.
    </div>
  </div>
</div>
`;

exports[`PendingUser should render rejected correctly with props and session 1`] = `
<div
  id="pendingUser"
>
  <div
    className="statusStatement"
  >
    Status: 
    rejected
  </div>
  <div
    className="progressStatement"
  >
    Progress: 
    2
    /
    7
  </div>
  <br />
  <div
    className="resultingStatement"
  >
    <div>
      All users have voted.
      <br />
      The required votes has not been met.
      <br />
      Your request has been rejected.
    </div>
  </div>
</div>
`;
